sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalysisContext.java

```diff
@@ -21,26 +21,44 @@
 
 package io.crate.analyze.expressions;
 
-import io.crate.analyze.symbol.Function;
-import io.crate.analyze.symbol.Symbol;
-import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.TransactionContext;
 import io.crate.sql.tree.ArrayComparisonExpression;
 import io.crate.sql.tree.Expression;
 
 import java.util.IdentityHashMap;
-import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 
+/**
+ * Base class for providing contextual information during expression analysis in
+ * {@link ExpressionAnalyzer}.
+ *
+ * 1) Function creation and folding constants (through symbol normalization),
+ *    e.g. turns '1 + 2' into 'Literal.of(3)'.
+ *
+ * 2) Detect sub queries in array expressions like '= ANY(Select id from t)'.
+ *
+ * 3) Indicates whether aggregates have been found during the analysis.
+ */
 public class ExpressionAnalysisContext {
 
     private final Map<Expression, Object> arrayExpressionsChildren = new IdentityHashMap<>();
 
-    public boolean hasAggregates = false;
+    private final TransactionContext transactionContext;
+
+    private boolean hasAggregates;
 
-    Function allocateFunction(FunctionInfo functionInfo, List<Symbol> arguments) {
-        Function newFunction = new Function(functionInfo, arguments);
-        hasAggregates = hasAggregates || functionInfo.type() == FunctionInfo.Type.AGGREGATE;
-        return newFunction;
+    public ExpressionAnalysisContext(TransactionContext transactionContext) {
```

Comment:
Does it still make sense to have the transactionContext here as well?

Reply:
We need the `TransactionContext` during normalization. I think it makes sense to store it in the context.

Reply:
Thought it might be an option to replace the SessionContext dependency in the ExpressionAnalyzer with the TransactionContext. In most cases the ExpressionAnalysisContext is created at the same place - so the dependency should be available there.